From 8ff171609d140fa89b63b5caae251a8a8eab671e Mon Sep 17 00:00:00 2001
From: Simeon Marijon <simeon.marijon@foss.st.com>
Date: Thu, 5 Oct 2023 19:00:02 +0200
Subject: [PATCH 0508/1695] drivers: nvmem: add write API for nvmem controllers

Add the ability to write to a nvmem cell

Upstream-Status: Inappropriate [ST downstream https://github.com/STMicroelectronics/optee_os.git 35248e2dac1fafa6a7bd995d386dfa104f8920c4 drivers: nvmem: add write API for nvmem controllers]
Signed-off-by: Simeon Marijon <simeon.marijon@foss.st.com>
Change-Id: I739501ddc54b5981bd717e2418b030bbc8ec15b6
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/331903
Domain-Review: Gatien CHEVALLIER <gatien.chevallier@st.com>
ACI: CIBUILD <MDG-smet-aci-builds@list.st.com>
Reviewed-by: Simeon MARIJON <simeon.marijon@st.com>
ACI: CITOOLS <MDG-smet-aci-reviews@list.st.com>
Tested-by: Simeon MARIJON <simeon.marijon@st.com>
Reviewed-by: Gatien CHEVALLIER <gatien.chevallier@st.com>
Reviewed-on: https://gerrit.st.com/c/mpu/oe/optee/optee_os/+/351773
---
 core/include/drivers/nvmem.h | 40 ++++++++++++++++++++++++++++++++++++
 1 file changed, 40 insertions(+)

diff --git a/core/include/drivers/nvmem.h b/core/include/drivers/nvmem.h
index 5db573d1f..c03d0950a 100644
--- a/core/include/drivers/nvmem.h
+++ b/core/include/drivers/nvmem.h
@@ -6,6 +6,7 @@
 #ifndef __DRIVERS_NVMEM_H
 #define __DRIVERS_NVMEM_H
 
+#include <assert.h>
 #include <kernel/dt_driver.h>
 #include <tee_api_defines.h>
 #include <tee_api_types.h>
@@ -27,6 +28,8 @@ struct nvmem_ops {
 	 * @data: Output buffer of size greater or equal to @cell->size
 	 */
 	TEE_Result (*read_cell)(struct nvmem_cell *cell, uint8_t *data);
+	TEE_Result (*write_cell)(struct nvmem_cell *cell, uint8_t *data,
+				 size_t len);
 	void (*put_cell)(struct nvmem_cell *cell);
 };
 
@@ -153,6 +156,22 @@ static inline TEE_Result nvmem_cell_read(struct nvmem_cell *cell,
 TEE_Result nvmem_cell_malloc_and_read(struct nvmem_cell *cell,
 				      uint8_t **out_data);
 
+/*
+ * nvmem_cell_write() - write data to a nvmem cell
+ * @cell: Cell to write
+ * @data: Input buffer
+ * @buf_len: Size of the buffer to write in the cell
+ */
+static inline TEE_Result nvmem_cell_write(struct nvmem_cell *cell,
+					  uint8_t *data, size_t buf_len)
+{
+	assert(buf_len <= cell->len);
+
+	if (!cell->ops->write_cell)
+		return TEE_ERROR_NOT_SUPPORTED;
+
+	return cell->ops->write_cell(cell, data, buf_len);
+}
 #else /* CFG_DRIVERS_NVMEM */
 static inline TEE_Result nvmem_register_provider(const void *fdt __unused,
 						 int nodeoffset __unused,
@@ -189,5 +208,26 @@ static inline TEE_Result nvmem_cell_parse_dt(const void *fdt __unused,
 static inline void nvmem_put_cell(struct nvmem_cell *cell __unused)
 {
 }
+
+static inline TEE_Result nvmem_cell_read(struct nvmem_cell *cell __unused,
+					 uint8_t *data __unused,
+					 size_t buf_len __unused,
+					 size_t *read_len __unused)
+{
+	return TEE_ERROR_NOT_SUPPORTED;
+}
+
+static inline TEE_Result nvmem_cell_write(struct nvmem_cell *cell __unused,
+					  uint8_t *data __unused,
+					  size_t buf_len __unused)
+{
+	return TEE_ERROR_NOT_SUPPORTED;
+}
+
+static inline TEE_Result nvmem_cell_malloc_and_read(struct nvmem_cell *cell,
+						    uint8_t **out_data)
+{
+	return TEE_ERROR_NOT_SUPPORTED;
+}
 #endif /* CFG_DRIVERS_NVMEM */
 #endif /* __DRIVERS_NVMEM_H */
-- 
2.47.2

